File: BooksApps Misc/QuickFileFilter Archives/QuickFileFilterVB/RelayCommand.vb
' RelayCommand
' Standard ViewModel helper for commanding, VB version
'
' 2016-11-28   PV

Friend Class RelayCommand(Of T) : Implements ICommand
    ReadOnly _canExecute As Predicate(Of T)
    ReadOnly _execute As Action(Of T)

    Public Sub New(execute As Action(Of T), canExecute As Predicate(Of T))
        _canExecute = canExecute
        _execute = execute
    End Sub

    ' canExecute Is optional, And by default Is assumed returning true (directly in CanExecute)
    Public Sub New(execute As Action(Of T))
        Me.New(execute, Nothing)
    End Sub

    ' From ICommand
    Public Function CanExecute(ByVal parameter As Object) As Boolean Implements ICommand.CanExecute
        Return If(_canExecute Is Nothing, True, _canExecute(CType(parameter, T)))
    End Function

    ' From ICommand
    Public Sub Execute(ByVal parameter As Object) Implements ICommand.Execute
        If _execute IsNot Nothing Then _execute(CType(parameter, T))
    End Sub


    ' The 'black magic' part: according to help, CommandManager.RequerySuggested Event occurs when the 
    ' CommandManager """detects conditions that might change the ability of a command to execute"""...
    ' Ok, it works, but exactly how does this detection works Is still a mystery to me...
    '
    ' Added info from CommandManager.InvalidateRequerySuggested Method:
    ' The CommandManager only pays attention to certain conditions in determining when the command target has changed, 
    ' such as change in keyboard focus. In situations where the CommandManager does Not sufficiently determine a change 
    ' in conditions that cause a command to Not be able to execute, InvalidateRequerySuggested can be called to force 
    ' the CommandManager to raise the RequerySuggested event. 

    ' From ICommand
    Public Custom Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged
        AddHandler(ByVal value As EventHandler)
            AddHandler CommandManager.RequerySuggested, value
        End AddHandler

        RemoveHandler(ByVal value As EventHandler)
            RemoveHandler CommandManager.RequerySuggested, value
        End RemoveHandler

        RaiseEvent()
            CommandManager.InvalidateRequerySuggested()
        End RaiseEvent
    End Event

End Class

File: Perceptive Hashing/DIF7/RelayCommand.cs
﻿// DIF - Duplicate Image Finder
// 2013-05-25   PV

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Input;

namespace DIF
{
    class RelayCommand<T> : ICommand
    {
        readonly Predicate<T> canExecute;
        readonly Action<T> execute;

        public RelayCommand(Action<T> execute, Predicate<T> canExecute)
        {
            this.canExecute = canExecute;
            this.execute = execute;
        }

        // canExecute is optional, and by default is assumed returning true (directly in CanExecute)
        public RelayCommand(Action<T> execute)
            : this(execute, null)
        { }

        /* From ICommand */
        public bool CanExecute(object parameter)
        {
            return canExecute == null ? true : canExecute((T)parameter);
        }

        /* From ICommand */
        public void Execute(object parameter)
        {
            if (execute != null) execute((T)parameter);
        }

        // The 'black magic' part: according to help, CommandManager.RequerySuggested Event occurs when the 
        // CommandManager """detects conditions that might change the ability of a command to execute"""...
        // Ok, it works, but exactly how does this detection works is still a mystery to me...
        //
        // Added info from CommandManager.InvalidateRequerySuggested Method:
        // The CommandManager only pays attention to certain conditions in determining when the command target has changed, 
        // such as change in keyboard focus. In situations where the CommandManager does not sufficiently determine a change 
        // in conditions that cause a command to not be able to execute, InvalidateRequerySuggested can be called to force 
        // the CommandManager to raise the RequerySuggested event. 

        /* From ICommand */
        public event EventHandler CanExecuteChanged
        {
            add { CommandManager.RequerySuggested += value; }
            remove { CommandManager.RequerySuggested -= value; }
        }
    }
}
